File: frontend/src/lib/api/librarian.ts
import { httpClient } from "@/lib/api/http";
import type { ApiMission } from "@/types/mission";

/** One message of the client-held transcript; resent on every call (decision #519). */
export interface TranscriptMessage {
  role: "user" | "assistant";
  content: string;
}

/**
 * One span of a Librarian reply. `prose` may use world knowledge freely;
 * `corpus_claim` is a statement about this project's evidence and carries
 * citations that resolve to `/evidence/{id}`; `withheld` is a claim the server
 * refused to render because its provenance did not hold.
 */
export interface ReplySegment {
  kind: "prose" | "corpus_claim" | "withheld";
  text: string;
  citations: string[];
}

export interface EvidenceRef {
  id: string;
  claim: string;
  source_url: string;
  disposition: string;
  href: string;
}

/** A document chunk an answer cites; `href` opens the document at that chunk (QA-1). */
export interface ChunkRef {
  id: string;
  document_id: string;
  document_name: string;
  chunk_index: number | null;
  snippet: string | null;
  href: string;
}

export interface TurnResponse {
  segments: ReplySegment[];
  suggested_action: "draft_mission" | null;
  evidence: EvidenceRef[];
  /** Chunks an answer turn cites; empty for a conversational turn. */
  chunks: ChunkRef[];
  /** True when an answer turn was refused: nothing in the project answers the question. */
  no_evidence: boolean;
  withheld_count: number;
  usage: Record<string, number> | null;
  model: string;
}

/**
 * The answer budgets, in tokens (decision #543). A citation label costs about 33
 * tokens, so 600 fits a short answer of about 150 words with its citations, and
 * 2000 a full synthesis of about 500. The server tells the model the length.
 */
export const ANSWER_BUDGETS = { short: 600, full: 2000 } as const;
export type AnswerBudget = keyof typeof ANSWER_BUDGETS;

export interface MissionDraft {
  mission_id: string;
  title: string;
  objective: string;
  success_criteria: string[];
  background?: string | null;
  focus?: string | null;
  required_entities?: string[] | null;
  excluded_entities?: string[] | null;
  constraints?: string[] | null;
  deliverable_format?: string | null;
  deliverables?: string[];
  tags?: string[];
}

export interface LintViolation {
  rule: string;
  field: string | null;
  message: string;
  suggestion: string | null;
}

export interface DraftPreview {
  contract_version: string;
  compiler_revision: string;
  fidelity: string;
  named_entities: string[];
  objectives: Array<Record<string, unknown>>;
  evidence_slots: Array<Record<string, unknown>>;
  acceptance_checks: Array<Record<string, unknown>>;
  deliverable_schemas: Array<Record<string, unknown>>;
  coverage_thresholds: Record<string, number>;
  validation_thresholds: Record<string, number>;
}

export interface DraftResponse {
  draft: MissionDraft;
  preview: DraftPreview | null;
  preview_error: string | null;
  lint_errors: LintViolation[];
  lint_warnings: LintViolation[];
  notes: string[];
  usage: Record<string, number> | null;
  model: string;
}

export interface CreatedMissionResponse {
  mission: ApiMission;
  created: boolean;
}

export const librarianApi = {
  /**
   * One turn; `projectId` scopes the evidence the Librarian may cite. With
   * `answer`, the last message is a question answered from the project's
   * documents within `maxTokens`, or refused.
   */
  turn(
    messages: TranscriptMessage[],
    projectId: string | null,
    answer?: { maxTokens: number },
  ): Promise<TurnResponse> {
    return httpClient.post<TurnResponse>(
      "/librarian/turns",
      answer
        ? { project_id: projectId, messages, mode: "answer", max_tokens: answer.maxTokens }
        : { project_id: projectId, messages },
    );
  },

  /** Turn the conversation into a compiled, linted mission draft. Nothing is saved. */
  draft(messages: TranscriptMessage[], projectId: string): Promise<DraftResponse> {
    return httpClient.post<DraftResponse>("/librarian/drafts", { project_id: projectId, messages });
  },

  /** Create the reviewed draft as a pristine draft mission. Explicit and idempotent. */
  createMission(draft: MissionDraft, projectId: string): Promise<CreatedMissionResponse> {
    return httpClient.post<CreatedMissionResponse>("/librarian/missions", { project_id: projectId, draft });
  },
};

/** Plain-text form of a reply, used when the transcript is resent to the server. */
export function replyToTranscriptText(segments: ReplySegment[], chunks: ChunkRef[] = []): string {
  const byId = new Map(chunks.map((chunk) => [chunk.id, chunk]));
  const cite = (id: string) => {
    const chunk = byId.get(id);
    return chunk ? `[document ${chunk.document_name}, chunk ${chunk.chunk_index ?? "?"}]` : `[evidence:${id}]`;
  };
  return segments
    .map((segment) =>
      segment.kind === "withheld"
        ? "[withheld claim]"
        : segment.citations.length
          ? `${segment.text} ${segment.citations.map(cite).join(" ")}`
          : segment.text,
    )
    .join("\n\n")
    .trim();
}

File: app/schemas/librarian.py
"""Request and response shapes for the Librarian (LIB-1).

The transcript is client-held and resent on every call: there is no
conversation table (decision #519). Every assistant reply is a list of typed
segments so the UI can render world knowledge and corpus claims differently,
which is the integrity guarantee decision #513 leaves load-bearing.
"""

from __future__ import annotations

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.mission import MissionBase, MissionResponse

TranscriptRole = Literal["user", "assistant"]
SegmentKind = Literal["prose", "corpus_claim", "withheld"]
SuggestedAction = Literal["draft_mission"]
TurnMode = Literal["converse", "answer"]

MAX_TRANSCRIPT_MESSAGES = 60
# The answer budget's bounds: RagQuery's floor, and twice the page's full synthesis.
ANSWER_MIN_TOKENS = 64
ANSWER_MAX_TOKENS = 4000


class TranscriptMessage(BaseModel):
    role: TranscriptRole
    content: str = Field(min_length=1, max_length=20_000)


class ReplySegment(BaseModel):
    """One span of an assistant reply.

    ``prose`` asserts nothing about the corpus and may use world knowledge freely.
    ``corpus_claim`` is a statement about what the project's evidence contains and
    must cite evidence entries retrieved in the same turn. ``withheld`` marks a
    claim the server refused to render because its provenance did not hold.
    """

    kind: SegmentKind = "prose"
    text: str = Field(default="", max_length=20_000)
    citations: list[str] = Field(default_factory=list, max_length=20)

    @field_validator("kind", mode="before")
    @classmethod
    def _normalize_kind(cls, value: Any) -> Any:
        if isinstance(value, str):
            lowered = value.strip().lower().replace("-", "_").replace(" ", "_")
            if lowered in {"claim", "corpus", "corpus_claim", "evidence"}:
                return "corpus_claim"
            if lowered in {"prose", "text", "general", "planning", "answer"}:
                return "prose"
            return lowered
        return value

    @field_validator("citations", mode="before")
    @classmethod
    def _coerce_citations(cls, value: Any) -> list[str]:
        if value is None:
            return []
        if isinstance(value, str):
            value = [value]
        if not isinstance(value, list):
            return []
        return [str(item).strip() for item in value if str(item).strip()][:20]


class AssistantReply(BaseModel):
    segments: list[ReplySegment] = Field(default_factory=list, max_length=50)
    suggested_action: SuggestedAction | None = None

    @field_validator("suggested_action", mode="before")
    @classmethod
    def _normalize_action(cls, value: Any) -> Any:
        if isinstance(value, str):
            lowered = value.strip().lower()
            return "draft_mission" if lowered in {"draft_mission", "draft", "mission"} else None
        return None


class TurnRequest(BaseModel):
    project_id: UUID | None = Field(
        default=None,
        description="Project whose evidence the Librarian may search. Omit for planning-only conversation.",
    )
    messages: list[TranscriptMessage] = Field(min_length=1, max_length=MAX_TRANSCRIPT_MESSAGES)
    mode: TurnMode = Field(
        default="converse",
        description=(
            "converse: the Librarian talks it through (LIB-1). answer: the last message is a question "
            "answered from the project's documents with citations, or refused (QA-1)."
        ),
    )
    max_tokens: int | None = Field(
        default=None,
        ge=ANSWER_MIN_TOKENS,
        le=ANSWER_MAX_TOKENS,
        description="Answer budget for mode 'answer'. Defaults to settings.rag_default_max_tokens.",
    )

    @model_validator(mode="after")
    def _last_message_is_from_user(self) -> TurnRequest:
        if self.messages[-1].role != "user":
            raise ValueError("The last transcript message must be from the user.")
        return self

    @model_validator(mode="after")
    def _answer_mode_fields(self) -> TurnRequest:
        if self.mode == "answer":
            if self.project_id is None:
                raise ValueError("Mode 'answer' needs a project_id: a question is answered from one project.")
            if not self.messages[-1].content.strip():
                raise ValueError("The question must not be empty.")
        elif self.max_tokens is not None:
            raise ValueError("max_tokens applies only to mode 'answer'.")
        return self


class EvidenceRef(BaseModel):
    """An evidence entry the Librarian consulted this turn, as the UI resolves it."""

    id: UUID
    claim: str
    source_url: str
    disposition: str
    href: str


class ChunkRef(BaseModel):
    """A document chunk an answer cites, as the UI resolves it (QA-1)."""

    id: str
    document_id: str
    document_name: str
    chunk_index: int | None = None
    snippet: str | None = None
    href: str


class TurnResponse(BaseModel):
    segments: list[ReplySegment]
    suggested_action: SuggestedAction | None = None
    evidence: list[EvidenceRef] = Field(default_factory=list)
    chunks: list[ChunkRef] = Field(default_factory=list)
    no_evidence: bool = Field(
        default=False,
        description="True when an answer turn was refused: nothing in the project answers the question.",
    )
    withheld_count: int = 0
    usage: dict[str, int] | None = None
    model: str


class MissionDraft(MissionBase):
    """The planning stage's output: the authoring fields a mission needs, nothing operational.

    Inherits mission_id, title, objective and success_criteria with the exact
    validators MissionCreate applies, so a draft that validates here creates
    unmodified.
    """

    background: str | None = None
    focus: str | None = None
    required_entities: list[str] | None = None
    excluded_entities: list[str] | None = None
    constraints: list[str] | None = None
    deliverable_format: str | None = None
    deliverables: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)

    @field_validator(
        "required_entities", "excluded_entities", "constraints", "deliverables", "tags", mode="before"
    )
    @classmethod
    def _clean_string_lists(cls, value: Any) -> Any:
        if value is None:
            return value
        if isinstance(value, str):
            value = [value]
        if isinstance(value, list):
            return [str(item).strip() for item in value if str(item).strip()]
        return value


class DraftRequest(BaseModel):
    project_id: UUID
    messages: list[TranscriptMessage] = Field(min_length=1, max_length=MAX_TRANSCRIPT_MESSAGES)


class LintViolationOut(BaseModel):
    rule: str
    field: str | None = None
    message: str
    suggestion: str | None = None


class DraftResponse(BaseModel):
    draft: MissionDraft
    preview: dict[str, Any] | None = None
    preview_error: str | None = None
    lint_errors: list[LintViolationOut] = Field(default_factory=list)
    lint_warnings: list[LintViolationOut] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)
    usage: dict[str, int] | None = None
    model: str


class CreateFromDraftRequest(BaseModel):
    project_id: UUID
    draft: MissionDraft


class CreatedMissionResponse(BaseModel):
    mission: MissionResponse
    created: bool
